### ISSUE TITLE
Randomly but consistently crashing during combat - BG

### ISSUE THREAD

**Issue by captain-cabinets** [exult/exult#257]
I'm maybe 1/4 of the way through BG, and haven't had any issues until now. Been to Yew and the Great Forest (or whatever it's called) before with no issues. Now I'm there and close to every time I start fighting something, Exult crashes. 

Stdout & Stderr attached. Save game attached (rename to .sav). If you load it up and walk NE to fight something near the wisps, it'll crash ~80% of the time. Or to the southwest where the agility trainer is with a couple of deer, it'll crash maybe 50% of the time when you try to attack the deer.

[stdout.txt](https://github.com/exult/exult/files/9941865/stdout.txt)
[stderr.txt](https://github.com/exult/exult/files/9941867/stderr.txt)
[exult14bg.txt](https://github.com/exult/exult/files/9941871/exult14bg.txt)

**Comment by captain-cabinets** [comment]
Update: It's not just in Yew. Seems to be the next instance of combat, wherever that is. I teleported to Britain, went west, got into a fight with a reaper, and it crashed straight away again.

**Comment by i30817** [comment]
Post the savegame.

**Comment by captain-cabinets** [comment]
I did, in the original post... but since github doesn't let you attach .sav files, I renamed it to a .txt. 

Here's a link as well: [https://drive.google.com/file/d/1B3cPYmdUXIHB843yVbYO5vS9WEtAyZak/view?usp=sharing]

**Comment by captain-cabinets** [comment]
[sorry, here](https://drive.google.com/file/d/1B3cPYmdUXIHB843yVbYO5vS9WEtAyZak/view?usp=sharing)

**Comment by rofl0r** [comment]
you can also attach zip files, that's what most people would do (put sav into zip, attach). either way you should probably try to get a backtrace with gdb when the crash happens, and recompile exult with memory sanitizer which should immediately point out invalid memory accesses.

**Comment by Dragon-Baroque** [comment]
Thank you for this report @captain-cabinets.

You might have wrapped the two logs and the savegame into a single zip. But you provided the logs and the savegame, this is the key.

From the stderr log, I obtained
```
terminate called after throwing an instance of 'std::bad_weak_ptr'
  what():  bad_weak_ptr
```
which points to memory corruption / bad smart pointer management.

From the stdout log, I obtained
```
Exult version 1.8.0·
Built at: Apr 14 2022 23:13:23
...
Platform: Windows 10 Version 10.0 Build 19044
```
which tells us that you are using Exult 1.8.0 on Windows 10.

With that, I confirm that I could reproduce your problem with your savegame against an 1.8.0 Exult running on Windows 10 and also on Linux.
But I could not reproduce it anymore against the latest 1.9.0git Exult either on Windows 10 or on Linux.

I rebuilt my local 1.8.0 with full debug and enabled memory sanitizer ( good idea @rofl0r ), and ran it under gdb :
It crashed consistently at the same code place, namely `Swap_weapons` in `combat.cc`, which was fixed in 1.9.0git by commit 2738a6b2885a7 from @marzojr, in response to the issue #211, which issue looks suspiciously similar to the present issue.

_Could you please download and try the latest snapshot of 1.9.x from `exult.info` ?_

**Comment by captain-cabinets** [comment]
Thanks to everyone, and thank you to @Dragon-Baroque for the investigation. Next time I'll do the smart thing and zip everything up together - unfortunately my brain only functions intermittently, and this was one of its off days :)

I can confirm that the latest 1.9.x snapshot fixes the issue and I have had no more crashes. So it does indeed seem to stem from the same issue as #211. I'll close this issue then. Thanks again for the help, I really appreciate it.
